etebase-server: use buildPythonApplication, default to withLdap = false

phaer opened the pull request:
* I noticed that using python3 here referred to pkgs.python3 while python is always the one from the calling interpreter. In other words: before that change
legacyPackages.x86_64-linux.python311.pkgs.etebase-server.python would give me a python 3.12 release, now it's 3.11 as it should be.

* as etebase-server is an app in pkgs/servers, not a library in pkgs/development/python-modules, we use buildPythonApplication and callPackage in all-packages.nix and convert with toPythonModule in python-packages.nix. The latter could be dropped in a future commit

* In services.etebase-server, we now use pkgs.etebase-server instead of python3.pkgs.etebase-server. We still expose passthru.python and passthru.python path in order to avoid rewriting the module right now as well. This could be simplified in a later commit, the wrapper there shouldn't be needed.

* withLdap defaults to false now, as pkgs.python3.pkgs.python-ldap is broken on master. (see e.g. https://github.com/NixOS/nixpkgs/pull/326321) and its not a required feature in any way.


## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

phaer committed:
etebase-server: use buildPythonApplication...

...not pythonPackage. This is an app, living in pkgs/servers, so
we just use callPackage in all-packages and drop the python module.

python3.pkgs.toPythonModule could be used if a python module was needed.
Before that change legacyPackages.x86_64-linux.python311.pkgs.etebase-server.python
would give a python 3.12 release, and was therefore broken.

etebase-server = toPythonModule (pkgs.etebase-server.override {
  python3 = python;
});

would now be correct

felschr commented:
Result of `nixpkgs-review pr 326398` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python311Packages.etebase-server</li>
    <li>python311Packages.etebase-server.dist</li>
  </ul>
</details>

felschr commented on pkgs/servers/etebase/default.nix:
```diff
@@ -55,7 +55,7 @@ python.pkgs.buildPythonPackage rec {
-  passthru.python = python;
+  passthru.python = python';
   # PYTHONPATH of all dependencies used by the package
   passthru.pythonPath = python.pkgs.makePythonPath propagatedBuildInputs;
```
```suggestion
  passthru.pythonPath = python'.pkgs.makePythonPath propagatedBuildInputs;
```

phaer replied:
Thanks! Not sure how i missed that with search & replace.

felschr reviewed (approved):
Looking good to me and tests pass on my machine with mentioned adjustment.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1821

Aleksanaa reviewed (dismissed):
This package almost works in an ugly way.

This is in `top-level/all-packages.nix`

```
etebase-server = with python3Packages; toPythonApplication etebase-server;
```

This is in `top-level/python-packages.nix`:

```
etebase-server = callPackage ../servers/etebase { };
```

I just don't know why we have to shoot our foot like this. The better way is just putting it in `by-name`, and if it is still used as a module, use `toPythonModule` in `python-packages.nix`. This PR only makes it more confusing to read, because normally people think of python2 when seeing `python` (because top-level `python` is python2)

phaer commented:
> This package almost works in an ugly way.

I agree with that. As it's an application it probably shouldn't be a python module as well, but I ofc don't know if there's any out of tree users of the module. The nixos module currently depends on python and pythonpath in passthru in the package, but that could of course be fixed.

>  This PR only makes it more confusing to read, because normally people think of python2 when seeing python (because top-level python is python2)

Hm, I was about to write that in the context of a python-module it makes sense. But yeah, the file isn't in pkgs/development/python-module, so thats a good argument to actually turn things around.

phaer committed:
etebase-server: default to withLdap = false

as it's not unsual to run it without ldap and python-ldap is currently broken on master for python 3.12.

felschr commented:
> I agree with that. As it's an application it probably shouldn't be a python module as well, [...]

The `etebase-server` NixOS module currently relies on the python module. If it could be refactored to work without it I would be happy removing the python module altogether.

It's been a while since I initially added the package & NixOS module, but I remember some issues why I couldn't get it working without the python module. The NixOS module invokes `etebase-server` via `uvicorn`, so that's a bit special.

I didn't know about `toPythonModule` yet, with that migrating to `pkgs/by-name` seems to be the best option to streamline the package.

felschr commented on pkgs/servers/etebase/default.nix:
```diff
@@ -1,6 +1,6 @@
 { lib
 , fetchFromGitHub
-, withLdap ? true
+, withLdap ? false
```
I guess this was just accidentally pushed? Or do you propose changing the default here?

phaer replied:
Yes, I do - see commit message. I'll update the PR description in a moment :)

felschr replied:
This should probably be mentioned in the release notes. And we might as well remove the python module then.

phaer replied:
Ah right, i just removed the python package, wrote 2 sentences of release notes, squashed and rebased.

Aleksanaa commented on nixos/doc/manual/release-notes/rl-2411.section.md:
```diff
@@ -206,6 +206,9 @@
 - The `services.mxisd` module has been removed as both [mxisd](https://github.com/kamax-matrix/mxisd) and [ma1sd](https://github.com/ma1uta/ma1sd) are not maintained any longer.
   Consequently the package `pkgs.ma1sd` has also been removed.
 
+- `python3.pkgs.etebase-server` has been removed, as it isn't shipped as python package by upstream and no users are known.
```
We don't have to add this

phaer replied:
Thanks for your review! Do you mean we *should* not add this? It seems to a relevant, user-visible change to effectively drop a package?

Aleksanaa commented on pkgs/top-level/python-packages.nix:
```diff
@@ -3974,8 +3974,6 @@ self: super: with self; {
     inherit (pkgs.darwin.apple_sdk.frameworks) Security;
   };
 
-  etebase-server = callPackage ../servers/etebase { };
```
But do remember to add an entry in python-aliases.nix!

phaer replied:
Can you explain why? We have an agreement between both maintainers of the package, @felschr and myself, to drop the python package.

SuperSandro2000 replied:
We always add an entry there except for extra versions. Then we can also drop the release notes entry.

phaer replied:
Ah right, sorry i misunderstood the suggestion at first. Should be fixed now?

SuperSandro2000 commented on pkgs/top-level/python-aliases.nix:
```diff
@@ -158,6 +158,7 @@ mapAliases ({
   email_validator = email-validator; # added 2022-06-22
   enhancements = throw "enhancements is unmaintained upstream and has therefore been removed"; # added 2023-10-27
   et_xmlfile = et-xmlfile; # added 2023-10-16
+  etebase-server = throw "pkgs.python3.etebase-server has been removed, use pkgs.etebase-server"; # added 2024-07-16
```
```suggestion
  etebase-server = pkgs.etebase-server; # added 2024-07-16
```
that should also just work

phaer replied:
You could end up with the wrong python version used, see this PRs description.
